File: sample/pascalast/bench/quicksort.pas
program QuickSortBench;
{ Repeated quicksort over a 4096-element array.
  Each round re-fills the array with a deterministic LCG sequence
  (so the sort actually does work each time).  Sums all entries
  after sorting and prints the running total — a result that
  depends on the data so the optimizer can't elide the work. }
const
  N      = 4096;
  ROUNDS = 500;
var
  i, r: integer;
  seed: integer;
  total: integer;
  a: array[1..4096] of integer;

procedure swap(i, j: integer);
var t: integer;
begin
  t := a[i]; a[i] := a[j]; a[j] := t
end;

procedure qsort(lo, hi: integer);
var
  pivot, i, j: integer;
begin
  if lo < hi then
  begin
    pivot := a[hi];
    i := lo - 1;
    for j := lo to hi - 1 do
      if a[j] <= pivot then
      begin
        i := i + 1;
        swap(i, j)
      end;
    swap(i + 1, hi);
    qsort(lo, i);
    qsort(i + 2, hi)
  end
end;

function lcg: integer;
begin
  seed := (seed * 1103515245 + 12345) mod 2147483648;
  lcg  := seed
end;

begin
  total := 0;
  seed := 1;
  for r := 1 to ROUNDS do
  begin
    for i := 1 to N do a[i] := lcg mod 100000;
    qsort(1, N);
    total := total + a[1] + a[N]
  end;
  writeln(total)
end.

File: sample/anpy/test/cases/quicksort.py
# In-place quicksort over a list of ints.
xs:[int] = None
i:int = 0

def qsort(a:[int], lo:int, hi:int) -> object:
    pivot:int = 0
    i:int = 0
    j:int = 0
    t:int = 0
    if lo >= hi:
        return None
    pivot = a[hi]
    i = lo
    j = lo
    while j < hi:
        if a[j] < pivot:
            t = a[i]
            a[i] = a[j]
            a[j] = t
            i = i + 1
        j = j + 1
    t = a[i]
    a[i] = a[hi]
    a[hi] = t
    qsort(a, lo, i - 1)
    qsort(a, i + 1, hi)
    return None

xs = [5, 2, 9, 1, 5, 6, 3, 8, 7, 0, 4]
qsort(xs, 0, len(xs) - 1)
i = 0
while i < len(xs):
    print(xs[i])
    i = i + 1
